src/Endemic/Repair.hs

```diff
@@ -90,42 +90,96 @@ getHoleFits ::
   plugRef <- initGhcCtxt' True cc local_exprs
-  -- Then we can actually run the program!
   setNoDefaulting
-  let exprFits expr =
-        liftIO (modifyIORef' plugRef resetHoleFitList)
-          >> handleSourceError
-            (getHoleFitsFromError plugRef)
-            (Right <$> compileParsedExpr expr)
-  mapM (\(l, e) -> mapLeft (zip l) <$> exprFits e) exprs
-    -- TODO: We get some refinement holefits here as well, but we just
-    -- throw them out currently
-    >>= processFits . map (map (second fst)) . lefts
+  res <- getHoleFits' plugRef (if holeLvl == 0 then 0 else holeDepth) exprs
+  return $ map (map (map snd . snd) . snd) res
   where
+    exprFits plugRef expr =
+      liftIO (modifyIORef' plugRef resetHoleFitList)
+        >> handleSourceError
+          (getHoleFitsFromError plugRef)
+          (Right <$> compileParsedExpr expr)
+    getHoleFits' ::
+      IORef HoleFitState ->
+      Int ->
+      [([SrcSpan], LHsExpr GhcPs)] ->
+      Ghc [(LHsExpr GhcPs, [(SrcSpan, [(Int, HsExpr GhcPs)])])]
+    getHoleFits' plugRef n exprs | n <= 0 = do
+      -- We don't want any more additional hole fits, so we just grab the
+      -- identifier ones.
+      -- We know we won't be lookning at the refinement hole-fits,
+      -- so we ignore them:
+      dflags <- getSessionDynFlags
+      setSessionDynFlags dflags {refLevelHoleFits = Just 0}
+      res <-
+        mapM (\(l, e) -> mapLeft ((e,) . zip l) <$> exprFits plugRef e) exprs
+          >>= processFits . map (second (map (second fst))) . lefts
+      dflags <- getSessionDynFlags
+      setSessionDynFlags dflags {refLevelHoleFits = Just 0}
+      return res
+    getHoleFits' plugRef n exprs = do
+      fits <- lefts <$> mapM (\(l, e) -> mapLeft ((e,) . zip l) <$> exprFits plugRef e) exprs
+      procs <- processFits $ map (second $ map (second (uncurry (++)))) fits
+      --       pprPanic "ppr" $ ppr procs
+      let f :: LHsExpr GhcPs -> SrcSpan -> [(Int, HsExpr GhcPs)] -> Ghc (SrcSpan, [(Int, HsExpr GhcPs)])
+          f expr loc fits = do
+            let (hasHoles, done) = partition ((> 0) . fst) fits
+                filled =
+                  mapMaybe
+                    ( \(lvl, fit) ->
+                        (fmap ((fit,) . first (replicate lvl)) . (`fillHole` expr)) fit
+                    )
+                    hasHoles
+
+            recur <- zip (map (L loc . fst) filled) . map snd <$> getHoleFits' plugRef (n -1) (map snd filled)
+            let repls = map (second (reverse . map (map snd . snd))) recur >>= uncurry replacements
+                procced :: [(Int, HsExpr GhcPs)]
+                procced = map ((0,) . unLoc . snd) repls
+            return (loc, done ++ procced)
+      mapM (\(e, xs) -> (e,) <$> mapM (uncurry (f e)) xs) procs
+
     mapLeft :: (a -> c) -> Either a b -> Either c b
     mapLeft f (Left a) = Left (f a)
     mapLeft _ (Right r) = Right r
 
 -- | Takes a list of list of list of hole fits and processes each fit so that
```

Comment:
I think here it makes some sense to say where it is used ? 

I guess this is the `RefinementHoleResult -> Fix` helper?

Reply:
It's in the where clause of the `getHoleFits` function, so the only place where this can be used is the `getHoleFits` function 😅